components/style/restyle_hints.rs

```diff
@@ -0,0 +1,247 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use selectors::Element;
+use selectors::matching::matches_compound_selector;
+use selectors::parser::{AttrSelector, Combinator, CompoundSelector, SimpleSelector};
+use selectors::states::*;
+use std::clone::Clone;
+use std::sync::Arc;
+use string_cache::{Atom, Namespace};
+
+/// When the ElementState of an element (like IN_HOVER_STATE) changes, certain
+/// pseudo-classes (like :hover) may require us to restyle that element, its
+/// siblings, and/or its descendants. Doing this conservatively is expensive,
+/// and so we RestyleHints to short-circuit work we know is unnecessary.
+///
+/// NB: We should extent restyle hints to check for attribute-dependent style
+/// in addition to state-dependent style (Gecko does this).
+
+
+bitflags! {
+    flags RestyleHint: u8 {
+        #[doc = "Rerun selector matching on the element."]
+        const RESTYLE_SELF = 0x01,
+        #[doc = "Rerun selector matching on all of the element's descendants."]
+        // NB: In Gecko, we have RESTYLE_SUBTREE which is inclusive of self, but heycam isn't aware
+        // of a good reason for that.
+        const RESTYLE_DESCENDANTS = 0x02,
+        #[doc = "Rerun selector matching on all later siblings of the element and all of their descendants."]
+        const RESTYLE_LATER_SIBLINGS = 0x04,
+    }
+}
+
+/// In order to compute restyle hints, we perform a selector match against a list of partial
+/// selectors whose rightmost simple selector may be sensitive to the thing being changed. We
+/// do this matching twice, once for the element as it exists now and once for the element as it
+/// existed at the time of the last restyle. If the results of the selector match differ, that means
+/// that the given partial selector is sensitive to the change, and we compute a restyle hint
+/// based on its combinator.
+///
+/// In order to run selector matching against the old element state, we generate a wrapper for
+/// the element which claims to have the old state. This is the ElementWrapper logic below.
+///
+/// Gecko does this differently for element states, and passes a mask called mStateMask, which
+/// indicates the states that need to be ignored during selector matching. This saves an ElementWrapper
+/// allocation and an additional selector match call at the expense of additional complexity inside
+/// the selector matching logic. This only works for boolean states though, so we still need to
+/// take the ElementWrapper approach for attribute-dependent style. So we do it the same both ways for
+/// now to reduce complexity, but it's worth measuring the performance impact (if any) of the
+/// mStateMask approach.
```

Comment:
@SimonSapin Just FYI here - after discussing with @bzbarsky, we decided to try using the ElementWrapper approach for state restyle hints, which may remove the need to have ElementState in rust-selectors (see the comment above). I think it'd be good to leave it there for the time being though, so that we can try it both ways at some point and measure.